src/node_wasi.cc

```diff
@@ -0,0 +1,1801 @@
+  wasi->memory_.Reset(wasi->env()->isolate(), args[0].As<Object>());
+}
+
+
+void WASI::readUInt8(char* memory, uint8_t* value, uint32_t offset) {
+  CHECK_NOT_NULL(memory);
+  CHECK_NOT_NULL(value);
+  *value = memory[offset] & 0xFF;
+}
+
+
+void WASI::readUInt16(char* memory, uint16_t* value, uint32_t offset) {
+  CHECK_NOT_NULL(memory);
+  CHECK_NOT_NULL(value);
+  *value = (memory[offset] & 0xFF) |
+           ((memory[offset + 1] & 0xFF) << 8);
+}
+
+
+void WASI::readUInt32(char* memory, uint32_t* value, uint32_t offset) {
+  CHECK_NOT_NULL(memory);
+  CHECK_NOT_NULL(value);
+  *value = (memory[offset] & 0xFF) |
+           ((memory[offset + 1] & 0xFF) << 8) |
+           ((memory[offset + 2] & 0xFF) << 16) |
+           ((memory[offset + 3] & 0xFF) << 24);
+}
+
+
+void WASI::readUInt64(char* memory, uint64_t* value, uint32_t offset) {
+  CHECK_NOT_NULL(memory);
+  CHECK_NOT_NULL(value);
+  uint64_t low = (memory[offset] & 0xFF) |
+                 ((memory[offset + 1] & 0xFF) << 8) |
+                 ((memory[offset + 2] & 0xFF) << 16) |
+                 ((memory[offset + 3] & 0xFF) << 24);
+  uint64_t high = (memory[offset + 4] & 0xFF) |
+                  ((memory[offset + 5] & 0xFF) << 8) |
+                  ((memory[offset + 6] & 0xFF) << 16) |
+                  ((memory[offset + 7] & 0xFF) << 24);
+  *value = (high << 32) + low;
+}
+
+
+void WASI::writeUInt8(char* memory, uint8_t value, uint32_t offset) {
+  CHECK_NOT_NULL(memory);
+  memory[offset] = value & 0xFF;
+}
+
+
+void WASI::writeUInt16(char* memory, uint16_t value, uint32_t offset) {
+  CHECK_NOT_NULL(memory);
+  memory[offset++] = value & 0xFF;
+  memory[offset] = (value >> 8) & 0xFF;
+}
+
+
+void WASI::writeUInt32(char* memory, uint32_t value, uint32_t offset) {
+  CHECK_NOT_NULL(memory);
+  memory[offset++] = value & 0xFF;
+  memory[offset++] = (value >> 8) & 0xFF;
+  memory[offset++] = (value >> 16) & 0xFF;
+  memory[offset] = (value >> 24) & 0xFF;
+}
+
+
+void WASI::writeUInt64(char* memory, uint64_t value, uint32_t offset) {
+  CHECK_NOT_NULL(memory);
+  memory[offset++] = value & 0xFF;
+  memory[offset++] = (value >> 8) & 0xFF;
+  memory[offset++] = (value >> 16) & 0xFF;
+  memory[offset++] = (value >> 24) & 0xFF;
+  memory[offset++] = (value >> 32) & 0xFF;
+  memory[offset++] = (value >> 40) & 0xFF;
+  memory[offset++] = (value >> 48) & 0xFF;
+  memory[offset] = (value >> 56) & 0xFF;
+}
+
+
+uvwasi_errno_t WASI::backingStore(char** store, size_t* byte_length) {
+  Environment* env = this->env();
+  Local<Object> memory = PersistentToLocal::Strong(this->memory_);
+  Local<Value> prop;
+
+  if (!memory->Get(env->context(), env->buffer_string()).ToLocal(&prop))
+    return UVWASI_EINVAL;
+
+  if (!prop->IsArrayBuffer())
+    return UVWASI_EINVAL;
+
+  Local<ArrayBuffer> ab = prop.As<ArrayBuffer>();
+  ArrayBuffer::Contents contents = ab->GetContents();
+  *byte_length = contents.ByteLength();
+  *store = static_cast<char*>(contents.Data());
```

Comment:
Heads up: You probably want to switch to `ab->GetBackingStore()` here, see e.g. https://github.com/addaleax/node/commit/e5964f94a53337350c428dc8c5c8d1e114b78a6b